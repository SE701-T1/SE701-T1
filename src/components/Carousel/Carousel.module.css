.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
}

.buttonContainer {
  display: flex;
  flex-direction: row;
  margin-top: 20px;
}

.buttonContainer > *:not(:first-child) {
  margin-left: 5px;
}

.circle {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  border: 1px solid #8896ac;
  cursor: pointer;
}

.active {
  background-color: #8896ac;
}
