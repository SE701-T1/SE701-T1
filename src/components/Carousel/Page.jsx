import React from 'react';
import PropTypes from 'prop-types';
import { Typography } from '@mui/material';
import styles from './Page.module.css';

export default function Page({ title, description }) {
    return (
        <>
            <Typography className={styles.title} variant='h5'>{title}</Typography>
            <Typography className={styles.description} variant='subtitle2'>{description}</Typography>
        </>
    )
}

Page.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
};