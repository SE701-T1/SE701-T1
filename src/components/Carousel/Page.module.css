.title {
  color: #292525;
}

.description {
  color: #666666;
}

@media screen and (min-width: 1024px) and (max-height: 1310px) {
  .title {
    font-size: 2.2rem;
  }

  .description {
    font-size: 1rem;
    margin-top: 2rem;
  }
}
