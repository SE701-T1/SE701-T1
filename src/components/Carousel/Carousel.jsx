import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Page from './Page';
import styles from './Carousel.module.css';

export default function Carousel({ data }) {
    const [activePage, setActivePage] = useState(data[0]);

    function onSetActivePage(index) {
        const page = data.find((value) => {
            if (value.id === index) {
                return true;
            }

            return false;
        });

        setActivePage(page);
    }

    return (
        <>
            <Page title={activePage.title} description={activePage.description}/>
            {
                data.map((item) =>
                    <div 
                        key={item.id} 
                        className={styles.circle} 
                        onClick={() => onSetActivePage(item.id)}
                        onKeyDown={() => onSetActivePage(item.id)}
                        role='button'
                        aria-label='Change page'
                        tabIndex={0}
                    />
                )
            }
        </>
    )
}

Carousel.propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired
    })).isRequired
};