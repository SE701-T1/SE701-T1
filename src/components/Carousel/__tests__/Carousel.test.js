import Carousel from "../Carousel";
import { fireEvent, render, screen } from "@testing-library/react";
import { act } from "react-test-renderer";
import renderer from 'react-test-renderer';

describe('Carousel', () => {
    const mockData = [
        {
            id: 1,
            title: 'page-one-title',
            description: 'page-one-description'
        },
        {
            id: 2,
            title: 'page-two-title',
            description: 'page-two-description'
        }
    ];

    beforeEach(() => {
        act(() => {
            render(<Carousel data={mockData}/>);
        });
    });

    it('renders two circle buttons in total', async () => {
        expect(await screen.findAllByRole('button')).toHaveLength(2);
    });

    it('initially renders the first page content, but not the subsequent pages', () => {
        expect(screen.getByText('page-one-title')).toBeInTheDocument();
        expect(screen.getByText('page-one-description')).toBeInTheDocument();
        expect(screen.queryByText('page-two-title')).not.toBeInTheDocument();
        expect(screen.queryByText('page-two-description')).not.toBeInTheDocument();
    });

    it('changes pages correct when the second page button is pressed', async () => {
        const pageButtons = await screen.findAllByRole('button');

        // This statement simulates a button click event on the second page button.
        fireEvent(pageButtons[1], new MouseEvent('click', {
            bubbles: true,
            cancelable: true
        }));

        expect(screen.queryByText('page-one-title')).not.toBeInTheDocument();
        expect(screen.queryByText('page-one-description')).not.toBeInTheDocument();
        expect(screen.getByText('page-two-title')).toBeInTheDocument();
        expect(screen.getByText('page-two-description')).toBeInTheDocument();
    });

    it('produces the correct initial snapshot', () => {
        const tree = renderer.create(<Carousel data={mockData}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
