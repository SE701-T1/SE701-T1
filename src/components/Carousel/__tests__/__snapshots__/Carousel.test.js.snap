// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel produces the correct initial snapshot 1`] = `
<div
  className="container undefined"
>
  <div>
    <h5
      className="MuiTypography-root MuiTypography-h5 title css-ag7rrr-MuiTypography-root"
    >
      page-one-title
    </h5>
    <h6
      className="MuiTypography-root MuiTypography-subtitle2 description css-dd2x5a-MuiTypography-root"
    >
      page-one-description
    </h6>
  </div>
  <div
    className="buttonContainer"
  >
    <div
      aria-label="Change page"
      className="circle active"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    />
    <div
      aria-label="Change page"
      className="circle "
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    />
  </div>
</div>
`;
